### PR TITLE
[SPARK-25857][core] Add developer documentation regarding delegation tokens.

### PR DESCRIPTION
**Pull request by vanzin**

### CODE REVIEW

**Comment by vanzin** [comment]
@tgravescs @jerryshao

**Review by HyukjinKwon — COMMENTED** [review]
Looks fine to me otherwise. I did a quick proof reading.

**Comment by HyukjinKwon** [comment]
It doesn't block this PR but how about considering to put this in somewhere like https://spark.apache.org/docs/latest/running-on-yarn.html after refining it? We have Kerberos chapter https://spark.apache.org/docs/latest/running-on-yarn.html#kerberos

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+Hadoop-to-local-OS-user mapping configured for the service. But the overall support should work
+for connecting to other services even when YARN is not being used.
+
+
+## Limitations of DT support in Spark
+
+There are certain limitations to bear in mind when talking about DTs in Spark.
+
+The first one is that not all DTs actually expose their renewal period. This is generally a
+service configuration that is not generally exposed to clients. For this reason, certain DT
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: `For this reason, certain DT
providers cannot provide a renewal period to the Spark code`
Just for the understanding are these HBase + Hive, right?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+that information, so in general it's a good idea for all services using DTs to use the same
+configuration as HDFS for the renewal period.
+
+The second one is that Spark doesn't always know what delegation tokens will be needed. For
+example, when submitting an application in cluster mode without a keytab, the launcher needs
+to create DTs without knowing what the application code will actually be doing. This means that
+Spark will try to get as many delegation tokens as is possible based on the configuration
+available. That means that if an HBase configuration is available to the launcher but the app
+doesn't actually use HBase, a DT will still be generated. The user would have to explicitly
+opt out of generating HBase tokens in that case.
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: Maybe worth to mention the provider enable flag?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+The third one is that it's hard to create DTs "as needed". Without access to Kerberos credentials,
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: `Without access to Kerberos credentials,
Spark cannot create DTs`
One exception is Kafka, which can get token with SSL certificate authentication.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+The above also implies that running impersonated applications in client mode can be a security
+concern, since arbitrary user code would have access to the same local content as the privileged
+user. But unlike token renewal, Spark does not prevent that configuration from running.
+
+When impersonating, the Spark launcher will create DTs for the "proxy" user. In the example
+used above, that means that when code authenticates to a service using the DTs, the authenticated
+user will be "B", not "A".
+
+Note that "proxy user" is a very Hadoop-specific concept. It does not apply to OS users (which
```
[the remark this answers is not in the document]
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Linux OS users and not kerberos users.

**Review by gaborgsomogyi — COMMENTED** [review]
High quality doc, thanks for creating it.
Found some minor stuff but looks good.

**Comment by gaborgsomogyi** [comment]
@HyukjinKwon do you mean here copy or link there?
`It doesn't block this PR but how about considering to put this in somewhere like https://spark.apache.org/docs/latest/running-on-yarn.html after refining it? We have Kerberos chapter https://spark.apache.org/docs/latest/running-on-yarn.html#kerberos`

**Comment by vanzin** [comment]
> how about considering to put this in somewhere like 

This is not for consumption by users. It's for consumption by Spark developers who want to mess with the delegation token functionality inside Spark.

I have to explain this stuff every time someone tries to use delegation tokens somewhere in Spark, so now I can just point to this doc.

**Comment by gaborgsomogyi** [comment]
> This is not for consumption by users. It's for consumption by Spark developers who want to mess with the delegation token functionality inside Spark.

+1 and this doc is not yarn specific.
I was thinking about a link maybe but started to think it's not good idea.

**Review by srowen — COMMENTED** [review]
I think this could reasonably go in the main Spark docs too, but up to you. I know it's long and fairly niche.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+This token can now be sent to other processes and used to authenticate to different daemons
+belonging to that service.
+
+And thus the first drawback of DTs becomes apparent: you need service-specific logic to create and
+use them. While it would be possible to create a shared API or even a shared service to manage the
+creation and use of DTs, that doesn't currently exist, and retrofitting such a system would be a
+huge change in a bunch of different services.
+
+Spark works around this by having a (somewhat) pluggable DT creation API. Support for new
+services can be added by implementing a "DT provider" that is then called by Spark when
```
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: This document is in the same package as the API. It's targeted at developers changing / using that API.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
+# Delegation Token Handling In Spark
+
+This document aims to explain and demistify delegation tokens as they are used by Spark, since
+this topic is generally a huge source of confusion.
+
+
+## What are delegation tokens?
+
+Delegation tokens (DTs from now on) are authentication tokens used by some services to replace
+Kerberos service tokens. Many services in the Hadoop ecosystem have support for DTs, since they
```
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I think listing all services that use DTs is kinda pointless. Otherwise, let's also includ Solr, Kudu, and who knows what else.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+Spark works around this by having a (somewhat) pluggable DT creation API. Support for new
+services can be added by implementing a "DT provider" that is then called by Spark when
+generating delegation tokens for an application. Spark distributes tokens to executors using
+the `UserGroupInformation` Hadoop API, and it's up to the DT provider and the respective
+client library to agree on how to use those tokens.
+
+Once they are created, the semantics of how DTs operate are also service-specific. But, in general,
+they try to follow the semantics of Kerberos tokens:
+
+* A "lifetime" which is for how long the DT is valid before it requires renewal.
```
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: There's an example in the next topic if this is not clear. This is the naming used by `kinit` (see its man page).

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 1 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: That's the case in the current providers available in Spark. But I can't say those are the only examples. This is not meant to explain the Spark code, it's meant to explain DTs in general.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: This is not user documentation, so not sure what benefit that would bring.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I see. How about adding the class name `HadoopDelegationTokenProvider` or explicitly mentioning it's an internal API? Maybe i'm too much caring but wonder if it could give an impression that we have an external API that support DT creation.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+part of application data.
+
+DTs allow for a single place (e.g. the Spark driver) to require Kerberos credentials. That entity
+can then distribute the DTs to other parts of the distributed application (e.g. Spark executors),
+so they can authenticate to services.
+
+* A single token is used for authentication
+
+If Kerberos authentication were used, each client connection to a server would require a trip
+to the Key Distribution Center (KDC) and generation of a service ticket. In a distributed system,
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: not a bit deal but I think we better stick to one way. " Key Distribution Center (KDC)" vs "TGT (ticket granting ticket)".

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Yes, I think we should differentiate renew and recreate, AFAIK in the Spark side we always create a new DT before the renewal interval.

**Comment by HyukjinKwon** [comment]
re: https://github.com/apache/spark/pull/23348#issuecomment-451229540

Yea, the current status looks closer to dev docs. But likewise I had to research and explain this stuff to users as well :-). Should be good to document in main doc somewhere after refining but yea let me leave it to you. Looks find to me anyway.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: The point here not the list but the fact there are services [which are] + [which aren't] in the Hadoop ecosystem. It's a minor thing so no strong opinion.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 1 above)
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Here I didn't expect any changes, just wanted to know which are fulfilling the criteria.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: The names are fine. AFAIK after lifetime reached the token can't be renewed but has to be recreated.

[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: > after lifetime reached the token can't be renewed 

No, that's the renewable life.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 8 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I'll just revert to KDC. Readers of this document are expected to at least be a little familiar with Kerberos, so those terms should be familiar.

**Comment by vanzin** [comment]
> explain this stuff to users as well 

It's fine to point users at this if you want. But it's not meant to be user documentation, because it doesn't provide anything actionable for users.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Yeah, written the wrong name. The point is such case worth to say recreate because renew is misleading.

**Review by gaborgsomogyi — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+Kerberos service tokens. Many services in the Hadoop ecosystem have support for DTs, since they
+have two very desirable advantages over Kerberos tokens:
+
+* No need to distribute Kerberos credentials
+
+In a distributed application, distributing Kerberos credentials is tricky. Not all users have
+keytabs, and when they do, it's generally frowned upon to distribute them over the network as
+part of application data.
+
+DTs allow for a single place (e.g. the Spark driver) to require Kerberos credentials. That entity
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: this isn't true, delegation tokens don't allow for single place.  you could just as well have a keytab in the single place that distributes the tgts. And really its not a single place since it distributes them to executors.  We might rephrase this as an implementation detail that they go to the driver and are distributed

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+
+* A single token is used for authentication
+
+If Kerberos authentication were used, each client connection to a server would require a trip
+to the KDC and generation of a service ticket. In a distributed system, the number of service
+tickets can balloon pretty quickly when you think about the number of client processes (e.g. Spark
+executors) vs. the number of service processes (e.g. HDFS DataNodes). That generates unnecessary
+extra load on the KDC, and may even run into usage limits set up by the KDC admin.
+
+
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: remove extra newline

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+And thus the first drawback of DTs becomes apparent: you need service-specific logic to create and
+use them. While it would be possible to create a shared API or even a shared service to manage the
+creation and use of DTs, that doesn't currently exist, and retrofitting such a system would be a
+huge change in a bunch of different services.
+
+Spark works around this by having a (somewhat) pluggable, internal DT creation API. Support for new
+services can be added by implementing a `HadoopDelegationTokenProvider` that is then called by Spark
+when generating delegation tokens for an application. Spark distributes tokens to executors using
+the `UserGroupInformation` Hadoop API, and it's up to the DT provider and the respective client
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: I think this is a yarn only way of distributing, @vanzin k8s, mesos doesn't use UGI does it?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: perhaps you should use max lifetime and expiration time, it has to be renewed before the expiration time but it can only be renewed up to the max lifetime

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+As discussed in the previous section, DTs are service-specific, and require service-specific
+libraries for creation *and* renewal. This means that for YARN to be able to renew application
+tokens, YARN needs:
+
+* The client libraries for all the services the application is using
+* Information about how to connect to the services the application is using
+* Permissions to connect to those services
+
+In reality, though, most of the time YARN has access to a single HDFS cluster, and that will be
+the extent of its DT renewal features. Any other tokens sent to YARN will be distributed to
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: I believe hive ones are also renewed.  I don't remember if hbase ones need renewed or are just a single lifetime. I think my point is this is a pretty broad statement that could be incorrect so I think we should change the wording.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+containers, but will not be renewed.
+
+This means that those tokens will expire way before their max lifetime, unless some other code
+takes care of renewing them.
+
+And, to make matters worse, not all client libraries even implement token renewal. To use the
+example of a service supported by Spark, the `renew()` method of HBase tokens is a no-op. So
+the only way to "renew" an HBase token is to create a new one.
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: while this is true I'm not sure this is an issue, perhaps we should just point out that different services implement them differently, not all of them need a renewal.    ie your first sentence "to make matters worse" I think should be removed.  In a way its nice that hbase tokens don't need renewal as you don't have to worry about that in your code.  But it does lead to inconsistency between services.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+
+## DT Renewal in Spark
+
+Because of the issues explained above, Spark implements a different way of doing renewal. Spark's
+solution is a compromise: it targets the lowest common denominator, which is services like HBase
+that do not support actual token renewal.
+
+In Spark, DT "renewal" is enabled by giving the application a Kerberos keytab. A keytab is
+basically your Kerberos password written into a plain text file, which is why it's so sensitive.
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: I think we should explicitly state the security ramifications of this.  distributed a keytab means that if its compromised someone can get tokens for any service as you.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+Delegation tokens (DTs from now on) are authentication tokens used by some services to replace
+Kerberos service tokens. Many services in the Hadoop ecosystem have support for DTs, since they
+have two very desirable advantages over Kerberos tokens:
+
+* No need to distribute Kerberos credentials
+
+In a distributed application, distributing Kerberos credentials is tricky. Not all users have
+keytabs, and when they do, it's generally frowned upon to distribute them over the network as
+part of application data.
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md]: We seem to be mixing the section of what is a delegation token with what the benefits are, the definition of what also seems to be in  the next section with the lifecycle, perhaps we need to reorganize this a bit to have more of a section on what is DT, within there you can go into the benefits or why

Might mention here that with tokens you can't acquire more tokens vs with tgts or keytab you can get more tokens so its more secure to use tokens.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: > recreate because renew is misleading

It's not. Here you *can* renew the token. The "renewable life" controls what you're saying.

> perhaps you should use max lifetime and expiration time

Again, this is the same naming as kerberos documentation will give you, which I think is, in the end, less confusing. I'll use HDFS names also, just in case.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 9 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: >  delegation tokens don't allow for single place

Hmm, yes they do. Yes, you could do those other things, but then you wouldn't need DTs. So DTs are one way to allow a single place to handle kerberos credentials.

>  delegation tokens don't allow for single place

Not sure I follow. Only a single place has *kerberos* credentials. DTs are distributed to executors, but they are not *kerberos* credentials.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+
+DTs allow for a single place (e.g. the Spark driver) to require Kerberos credentials. That entity
+can then distribute the DTs to other parts of the distributed application (e.g. Spark executors),
+so they can authenticate to services.
+
+* A single token is used for authentication
```
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: The intent here is explained in the following paragraph. I'll add "per service" but I don't see how any other rephrasing can replace the following paragraph.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+(ticket granting ticket), which can then be used to request service tickets.
+
+There are other ways of obtaining TGTs, but, ultimately, you need a TGT to bootstrap the process.
+
+Once a TGT is available, the target service's client library can then be used to authenticate
+to the service using the Kerberos credentials, and request the creation of a delegation token.
+This token can now be sent to other processes and used to authenticate to different daemons
+belonging to that service.
+
+And thus the first drawback of DTs becomes apparent: you need service-specific logic to create and
```
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: It explains why Spark needs so much code to deal with delegation tokens, including an extensible interface, that wouldn't be needed if they were more like kerberos (where you just talk to one place).

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 11 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: You're thinking about writing the tokens to a file and automatically loading them, which is a YARN-only thing. Spark has its own mechanism to distribute tokens, which it uses even in YARN.

The YARN token API is mostly used so YARN can use the app's tokens for log aggregation and stuff.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 12 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: YARN does not have Hive client libraries in its classpath, so how can it even talk to the HMS or HS2 at all?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 13 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: This is build-up for explaining why Spark handles DTs the way it does.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 14 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Explain more than "it's like writing your password to a file"?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 12 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok checked with our hive team I was wrong.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 11 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: I'm not following the sentence above says spark distributes tokens to the executors using the UGI hadoop api. this works on yarn because its passed when you launch an executor, but how does the UGI get from the driver to executors on mesos?

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 9 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Oh I think I see what you are trying to say here.  I think you need to elaborate more on this.  I think you are trying to say that if you are sending kerberos credentials like a keytab that allows us to do all DT handling in a single place - the driver.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok, I guess I don't see the benefit in talking about what we could do, I would prefer to see it just state the drawback and if you want to compare to kerberos, compare to it.  But its not that big of a deal either.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 14 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Yes, I can see people not understanding that is a security concern.   If the file is written to HDFS with permissions restrictive, that might seem ok, but many security people would disagree.  Perhaps just add another sentence talking about this is a security concern because if the keytab is compromised people can get tokens for any service for an indefinite amount of time.  So users should evaluate if using this approach is ok.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 9 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: No, what I'm trying to say is that when you use DTs, there's a single place that needs to handle kerberos credentials.

It doesn't matter if you're using keytabs or kinit. It's still a single place. And that single place is responsible for distributing the DTs to the other entities that needs to authenticate against services.

It also doesn't need to be the Spark driver. As I mention in other places in the document, the DTs can be created by an external entity (like Oozie does).

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok, I'll remove.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 11 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: Spark distributes the tokens using its RPC system, but they're stashed in the UGI, so if an executor needs to get the tokens, it needs to look at the UGI.

Anyway, I'll replace "distributes" since it seems to be causing confusion.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 14 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: It sounds like you want to warn users of how dangerous this can be. This is not end-user documentation, so even if I make that statement here, it will probably not have the effect you want.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md (hunk 9 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md]: ok, sorry I understand now, I somehow really misread this.

**Comment by tgravescs** [comment]
one minor comment, otherwise +1, sorry again about my confusion yesterday on the one section.

**Comment by tgravescs** [comment]
+1

**Comment by vanzin** [comment]
Merging to master.

File: core/src/main/scala/org/apache/spark/deploy/security/README.md [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+not generate tokens, depending on what makes more sense in the context.
+
+
+## Externally Generated DTs
+
+Spark uses the `UserGroupInformation` API to manage the Hadoop credentials. That means that Spark
+inherits the feature of loading DTs automatically from a file. The Hadoop classes will load the
+token cache pointed at by the `HADOOP_TOKEN_FILE_LOCATION` environment variable, when it's defined.
+
+In this situation, Spark will not create DTs for the services that already have tokens in the
```
Comment (gaborgsomogyi) [inline core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: @tgravescs Since not many peoples are dealing with security maybe just you and me so asking you directly. I've just double checked the code because wanted to use the mentioned feature but I have the feeling it's not working. After some digging I've not seen any significant code changes in the security area so I have the feeling it's maybe never worked. Have you ever looked after this feature? When I take a look at the current code I see that the obtained tokens are overwriting the existing ones.

Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: As far as I can remember this worked at one point and time, but there has been a lot of changes to the security code since I possibly tried it. I don't really deal with security anymore so don't have a setup to try.

Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: Good to hear at least once it was working. Do you know by heart roughly where to look for the mentioned filtering?
I though I would see some if condition not to set UGI but there is no such code.

Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: Sounds like the problem is that it picks them up by then overwrites?  I would look for the calls to addCredentials, I don't remember the details on that api to know if they overwrite or merge things. Seems like at one point we specifically handled merging but like I said its been a long time since I wrote that code and many people changed it since.

Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: The HADOOP_TOKEN_FILE_LOCATION stuff was just all handled by Hadoop for us and we inherit so might not see any code directly related to that. If its overwriting them, my guess is we are just smashing in the tokens rather then dealing with them there.

Comment (gaborgsomogyi) [reply core/src/main/scala/org/apache/spark/deploy/security/README.md:214]: Thx for the pointers, going to sit together with the hadoop guys and having a look at HADOOP_TOKEN_FILE_LOCATION + addCredentials in depth. AFIAR addCredentials overwrites by default but will see... Thx again!
